src/FanoutGraphqlExpressServer.ts

```diff
@@ -0,0 +1,629 @@
+      console.log("connectionListener#onMessage", connection.id, message);
+      events.emit("message", message);
+      return;
+    },
+    async onOpen() {
+      console.log("connectionListener#onOpen", connection.id);
+      socket.readyState = SocketReadyState.OPEN;
+      return;
+    },
+  };
+};
+
+interface IGraphqlWebSocketOverHttpConnectionListenerOptions {
+  /** Info about the WebSocket-Over-HTTP Connection */
+  connection: IWebSocketOverHTTPConnectionInfo;
+  /** Grip configuration */
+  grip: {};
+  /** Handle a websocket message and optionally return a response */
+  getMessageResponse(message: string): void | string | Promise<string | void>;
+}
+
+/**
+ * Connection Listener that tries to mock out a basic graphql-ws.
+ * It's also grip and WebSocket-Over-HTTP aware.
+ */
+const GraphqlWebSocketOverHttpConnectionListener = (
+  options: IGraphqlWebSocketOverHttpConnectionListenerOptions,
+): IConnectionListener => {
+  return {
+    async onMessage(message) {
+      const graphqlWsEvent = JSON.parse(message);
+      if (graphqlWsEvent.type === "start") {
+        const query = gql`
+          ${graphqlWsEvent.payload.query}
+        `;
+        const mainDefinition = getMainDefinition(query);
+        const selections = mainDefinition.selectionSet.selections;
+        const selection = selections[0];
+        if (!selection) {
+          throw new Error("could not parse selection from graphqlWsEvent");
+        }
+        if (selection.kind === "Field") {
+          const selectedFieldName = selection.name.value;
+          const gripChannel = selectedFieldName;
+          console.debug(
+            `GraphqlWebSocketOverHttpConnectionListener requesting grip subscribe to channel ${gripChannel}`,
+          );
+          options.connection.webSocketContext.subscribe(gripChannel);
+        }
+      }
+      return options.getMessageResponse(message);
+    },
+  };
+};
+
+interface IGripOptions {
+  /** Grip Control URL */
+  url: string;
+}
+
+interface IGripPubSubOptions {
+  /** grip options */
+  grip: IGripOptions;
+}
+
+const GripPubSub = (
+  pubsub: PubSubEngine,
+  subscriptionType: GraphQLObjectType,
+  options: IGripPubSubOptions,
+): PubSubEngine => {
+  const gripPubControl = new grip.GripPubControl({
+    control_uri: options.grip.url,
+  });
+  const createGraphqlWsMessageForPublish = (
+    triggerName: string,
+    payload: any,
+  ) => {
+    const fieldForTrigger = subscriptionType.getFields()[triggerName];
+    if (fieldForTrigger) {
+      const fieldReturnTypeName = (() => {
+        const fieldType = fieldForTrigger.type;
+        if ("name" in fieldType) {
+          return fieldType.name;
+        }
+        if ("ofType" in fieldType) {
+          // e.g. a NotNullType
+          return fieldType.ofType.name;
+        }
+        assertNever(fieldType);
+      })();
+      return JSON.stringify({
+        id: "1", // TODO: this should be based on the subscription's graphqlWsEvent.id
```

Comment:
Note this TODO. Implementing would require being able to look up subscription information somewhere.